kmod/src/data.c

```diff
@@ -1565,48 +1563,95 @@ int scoutfs_data_fiemap(struct inode *inode, struct fiemap_extent_info *fieinfo,
 	iblock = start >> SCOUTFS_BLOCK_SM_SHIFT;
 	last = (start + len - 1) >> SCOUTFS_BLOCK_SM_SHIFT;
 
+	args.ino = ino;
+	args.inode = inode;
+
+	/* outer loop */
 	while (iblock <= last) {
-		ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
-				       iblock, 1, &ext);
-		if (ret < 0) {
-			if (ret == -ENOENT)
-				ret = 0;
-			last_flags = FIEMAP_EXTENT_LAST;
+		/* lock */
+		inode_lock(inode);
+		down_read(&si->extent_sem);
+
+		ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_READ, 0, inode, &lock);
+		if (ret) {
+			lock = NULL;
 			break;
 		}
 
-		trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+		args.lock = lock;
 
-		if (ext.start > last) {
-			/* not setting _LAST, it's for end of file */
-			ret = 0;
-			break;
+		/* collect entries */
+		info = page_address(page);
+		while (entries < (PAGE_SIZE / sizeof(struct fiemap_extent)) - 1) {
+			ret = scoutfs_ext_next(sb, &data_ext_ops, &args,
+					       iblock, 1, &ext);
+			if (ret < 0) {
+				if (ret == -ENOENT)
+					ret = 0;
+				last_flags = FIEMAP_EXTENT_LAST;
+				break;
+			}
+
+			trace_scoutfs_data_fiemap_extent(sb, ino, &ext);
+
+			if (ext.start > last) {
+				/* not setting _LAST, it's for end of file */
+				ret = 0;
+				break;
+			}
+
+			if (scoutfs_ext_can_merge(&cur, &ext)) {
+				/* merged extents could be greater than input len */
+				cur.len += ext.len;
+			} else {
+				/* fill it */
+				info->fe_logical = cur.start;
+				info->fe_length = cur.len;
+				info->fe_physical = cur.map;
+				info->fe_flags = cur.flags;
+
+				entries++;
+				info++;
+
+				cur = ext;
+			}
+
+			iblock = ext.start + ext.len;
 		}
 
-		if (scoutfs_ext_can_merge(&cur, &ext)) {
-			/* merged extents could be greater than input len */
-			cur.len += ext.len;
-		} else {
-			ret = fill_extent(fieinfo, &cur, 0);
+		/* unlock */
+		if (lock)
+			scoutfs_unlock(sb, lock, SCOUTFS_LOCK_READ);
+		up_read(&si->extent_sem);
+		inode_unlock(inode);
+
+		if (ret)
+			break;
+
+		/* emit entries */
+		info = page_address(page);
+		for (; entries > 0; entries--) {
+			ext.start = info->fe_logical;
+			ext.len = info->fe_length;
+			ext.map = info->fe_physical;
+			ext.flags = info->fe_flags;
+
+			ret = fill_extent(fieinfo, &ext, 0);
 			if (ret != 0)
-				goto unlock;
-			cur = ext;
+				goto out;
+			info++;
 		}
 
-		iblock = ext.start + ext.len;
+		/* still one left, it's in cur */
+		if (last_flags) {
+			ret = fill_extent(fieinfo, &cur, last_flags);
+			goto out;
+		}
 	}
 
-	if (cur.len)
-		ret = fill_extent(fieinfo, &cur, last_flags);
-unlock:
-	scoutfs_unlock(sb, lock, SCOUTFS_LOCK_READ);
-	up_read(&si->extent_sem);
-	inode_unlock(inode);
-
 out:
-	if (ret == 1)
```

Comment:
I'm not sure why this was removed.  It's here 'cause fill_extent can return 1 when it should stop filling and return and we didn't want to let that escape our inner loop.

Reply:
Overlooked this and it'll be put back.